File: 2021/6/cpp/main-b.cpp
#include <iostream>
#include <fstream>
#include <string_view>
#include <string>
#include <filesystem>
namespace fs = std::filesystem;


#include <unordered_map>


std::vector<std::string> split(const std::string& s, const std::string& delim = " ")
{
  std::vector<std::string> tokens;

  size_t start = 0;
  size_t end = 0;
  while((end = s.find(delim, start)) != std::string::npos)
  {
    tokens.emplace_back(s.substr(start, end-start));
    start = end + delim.size(); 
  }
  tokens.emplace_back(s.substr(start));

  return tokens;
}

void solve(const std::vector<std::string>& rawInput)
{
  // std::cout << "Printing the input inside solution():" << std::endl;
  // for (auto line : rawInput) {
  //   std::cout << "Line: " << line << std::endl;
  // }

  const size_t MAX_ITERATIONS = 256;
  size_t i = 0;

  std::unordered_map<int, long long> fish;
  for (size_t i = 0; i < 9; ++i)
    fish[i] = 0;

  const std::vector<std::string> input = split(rawInput[0], ",");
  for (const std::string& e : input)
  {
    const int nr = std::stoi(e);
    fish[nr]++;
  }
  while (i < MAX_ITERATIONS)
  {
    ++i;
    const long long fish0 = fish[0];

    for (size_t j = 0; j < 9-1; ++j)
    {
      fish[j] = fish[j+1];
    }
    fish[8] = fish0;
    fish[6] += fish0;

    // std::cout << "After " << i << " days (size=" << fish.size() <<"): ";
    // for (const int& f : fish)
    //   std::cout << f << ",";
    // std::cout << std::endl;
  }

  long long count = 0;
  for (size_t i = 0; i < 9; ++i)
    count += fish[i];
  std::cout << "Final count: " << count << std::endl;
}

void run(std::string_view filename)
{
  // Check if the file exists.
  const fs::path filepath { filename };
  if (!fs::exists(filepath))
  {
    std::cout << "File " << filepath.string() << " does not exist!" << std::endl;
    return;
  }

  std::cout << "Running the program using the input file: " << filepath.string() << std::endl;

  // Read the raw input and pass it in.
  std::vector<std::string> rawInput;

  std::ifstream file(filepath, std::ios::in);
  for(std::string line; std::getline(file, line); )
  {
    rawInput.emplace_back(line);
  }

  solve(rawInput);
}

int main(int argc, char** argv)
{
  // Parse arguments.
  const char* INPUT_FILENAME = { "input.txt" };
  std::string_view filename;
  if (argc < 2)
  {
    // std::cout << "Missing filename argument!" << std::endl;
    filename = INPUT_FILENAME;
  }
  else
  {
    filename = argv[1];
  }

  // Run the program.
  run(filename);

  return 0;
}

File: 2021/6/py/main-b.py
import argparse

#from dataclasses import dataclass

def solve(raw_input):
  MAX_ITERATIONS = 256
  fish = {}
  for i in range(9):
    fish[i] = 0

  input = [int(x) for x in raw_input.split(',')]

  for i in input:
    fish[i] += 1

  iter = 0
  while iter < MAX_ITERATIONS:
    iter += 1
    fish0 = fish[0]

    for i in range(8):
      fish[i] = fish[i+1]

    fish[8] = fish0
    fish[6] += fish0

  count = 0
  for i in range(9):
    count += fish[i]

  print(f"{count=}")


def main(filename):
  # Read file and convert to ints
  input = []
  with open(filename) as file:
    input = file.read()
    solve(input)

if __name__ == '__main__':

  # Simple argparse for the input
  parser = argparse.ArgumentParser()
  parser.add_argument("filename", type=str, nargs="?", default="input.txt",
        help="Enter the name of the file (default = input.txt)")
  parser.add_argument("-i", "--interactive", default="false", action="store_true", dest="interactive",
        help="If flag is set, the input is interactive")
  args = parser.parse_args()

  # Run the program
  main(args.filename)